File: tests/.hooks/php/primary/src/Hooks/TestHook.php
<?php

declare(strict_types=1);

namespace OpenAPI\OpenAPI\Hooks;

use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Promise\PromiseInterface;
use OpenAPI\OpenAPI\SDKConfiguration;
use OpenAPI\OpenAPI\Utils;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\UriInterface;


class TestClient implements \GuzzleHttp\ClientInterface
{
    public \GuzzleHttp\ClientInterface $client;

    public function __construct(?\GuzzleHttp\ClientInterface $client = null)
    {
        if ($client === null) {
            $client = new \GuzzleHttp\Client();
        } else {
            $this->client = $client;
        }
    }

    /**
     * Send an HTTP request.
     *
     * @param  RequestInterface  $request  Request to send
     * @param  array<string, mixed>  $options  Request options to apply to the given
     *                                         request and to the transfer.
     *
     * @throws GuzzleException
     */
    public function send(RequestInterface $request, array $options = []): ResponseInterface
    {
        $options['headers']['Client-Level-Header'] = 'added by client';

        return $this->client->send($request, $options);
    }


    /**
     * Asynchronously send an HTTP request.
     *
     * @param  RequestInterface  $request  Request to send
     * @param  array<string, mixed>  $options  Request options to apply to the given
     *                                         request and to the transfer.
     */
    public function sendAsync(RequestInterface $request, array $options = []): PromiseInterface
    {
        $options['headers']['Client-Level-Header'] = 'added by client';

        return $this->client->sendAsync($request, $options);
    }

    /**
     * Get a client configuration option.
     *
     * These options include default request options of the client, a "handler"
     * (if utilized by the concrete client), and a "base_uri" if utilized by
     * the concrete client.
     *
     * @param  string|null  $option  The config option to retrieve.
     *
     * @return mixed
     *
     * @deprecated ClientInterface::getConfig will be removed in guzzlehttp/guzzle:8.0.
     */
    public function getConfig(?string $option = null): mixed
    {
        return $this->client->getConfig($option);
    }

    /**
     * Create and send an HTTP request.
     *
     * Use an absolute path to override the base path of the client, or a
     * relative path to append to the base path of the client. The URL can
     * contain the query string as well.
     *
     * @param  string  $method  HTTP method.
     * @param  string|UriInterface  $uri  URI object or string.
     * @param  array<string, mixed>  $options  Request options to apply.
     *
     * @throws GuzzleException
     */
    public function request(string $method, $uri, array $options = []): ResponseInterface
    {
        $options['headers']['Client-Level-Header'] = 'added by client';

        return $this->client->request($method, $uri, $options);
    }

    /**
     * Create and send an asynchronous HTTP request.
     *
     * Use an absolute path to override the base path of the client, or a
     * relative path to append to the base path of the client. The URL can
     * contain the query string as well. Use an array to provide a URL
     * template and additional variables to use in the URL template expansion.
     *
     * @param  string  $method  HTTP method
     * @param  string|UriInterface  $uri  URI object or string.
     * @param  array<string, mixed>  $options  Request options to apply.
     */
    public function requestAsync(string $method, $uri, array $options = []): PromiseInterface
    {
        $options['headers']['Client-Level-Header'] = 'added by client';

        return $this->client->requestAsync($method, $uri, $options);
    }
}


class TestHook implements AfterErrorHook, AfterSuccessHook, BeforeRequestHook, SDKInitHook
{
    private string $initSdkVersion = '';

    public function sdkInit(SDKConfiguration $config): SDKConfiguration
    {
        $this->initSdkVersion = $config->sdkVersion;
        $config->client = new TestClient($config->client);

        return $config;
    }

    public function beforeRequest(BeforeRequestContext $context, RequestInterface $request): RequestInterface
    {
        $request = $request->withHeader('Idempotency-Key', 'some-key');

        switch ($context->operationID) {
            case 'testHooks':
                $requestParams = Utils\Utils::proper_parse_str($request->getUri()->getQuery());
                $requestParams['someParam'] = 'overriddenParam';
                $uri = $request->getUri()->withQuery(http_build_query($requestParams));
                $request = $request->withUri($uri);
                break;
            case 'authorizationHeaderModification':
                $token = $context->securitySource->call($context)->apiKeyAuth;
                $request = $request->withoutHeader('Authorization')->withAddedHeader('Authorization', "$token modified");
                break;
            case 'testHooksBeforeCreateRequestPaths':
                $request = $request->withAddedHeader('old-pathname', $request->getUri()->getPath());
                break;
            case 'hooksCustomUserAgent':
                $request = $request
                    ->withHeader('user-agent', 'acme-corp/'.$context->config->sdkVersion.' acme-corp/'.PHP_VERSION)
                    ->withHeader('X-Test-Gen-Version', $context->config->genVersion)
                    ->withHeader('X-Test-Doc-Version', $context->config->openapiDocVersion)
                    ->withHeader('X-Test-Init-Sdk-Version', $this->initSdkVersion);
                break;
        }

        return $request;
    }

    public function afterSuccess(AfterSuccessContext $context, ResponseInterface $response): ResponseInterface
    {
        if ($context->operationID === 'testHooksAfterResponse') {
            throw new \Exception('validation failed');
        }

        return $response;
    }

    public function afterError(AfterErrorContext $context, ?ResponseInterface $response, ?\Throwable $exception): ErrorResponseContext
    {
        if ($context->operationID == 'testHooksError') {
            if ($response != null && $response->getStatusCode() !== 400) {
                return new ErrorResponseContext(null, new \Exception('expected status code 400'));
            }

            return new ErrorResponseContext(null, new \Exception('special test error case'));
        }

        if ($context->operationID == 'statusGetDefaultError') {
            if ($response != null && $response->getStatusCode() === 418) {
                $okResponse = new \GuzzleHttp\Psr7\Response(200);

                return new ErrorResponseContext($okResponse, null);
            }
        }

        return new ErrorResponseContext($response, $exception);
    }
}

File: tests/.hooks/go/primary/internal/hooks/testhook.go
package hooks

import (
	"errors"
	"fmt"
	"net/http"
	"runtime"

	"example.com/openapi-go-sdk/internal/config"
)

type TestHook struct {
	initSDKVersion string
}

var (
	_ sdkInitHook       = (*TestHook)(nil)
	_ beforeRequestHook = (*TestHook)(nil)
	_ afterSuccessHook  = (*TestHook)(nil)
	_ afterErrorHook    = (*TestHook)(nil)
)

func (i *TestHook) SDKInit(config config.SDKConfiguration) config.SDKConfiguration {
	i.initSDKVersion = config.SDKVersion

	client := config.Client

	if client == nil {
		client = &http.Client{}
	}

	config.Client = &TestClient{
		underlyingClient: client,
	}

	return config
}

func (i *TestHook) BeforeRequest(hookCtx BeforeRequestContext, req *http.Request) (*http.Request, error) {
	req.Header.Set("Idempotency-Key", "some-key")

	switch hookCtx.OperationID {
	case "testHooks":
		values := req.URL.Query()
		values.Set("someParam", "overriddenParam")
		req.URL.RawQuery = values.Encode()
	case "authorizationHeaderModification":
		req.Header.Set("Authorization", req.Header.Get("Authorization")+" modified")
	case "hooksCustomUserAgent":
		req.Header.Set("User-Agent", fmt.Sprintf("acme-corp/%s acme-corp/%s", hookCtx.SDKConfiguration.SDKVersion, runtime.Version()))
		req.Header.Set("X-Test-Gen-Version", hookCtx.SDKConfiguration.GenVersion)
		req.Header.Set("X-Test-Doc-Version", hookCtx.SDKConfiguration.OpenAPIDocVersion)
		req.Header.Set("X-Test-Init-Sdk-Version", i.initSDKVersion)
	}

	return req, nil
}

func (i *TestHook) AfterSuccess(hookCtx AfterSuccessContext, res *http.Response) (*http.Response, error) {
	switch hookCtx.OperationID {
	case "testHooksAfterResponse":
		return nil, errors.New("validation failed")
	}

	return res, nil
}

func (i *TestHook) AfterError(hookCtx AfterErrorContext, res *http.Response, err error) (*http.Response, error) {
	switch hookCtx.OperationID {
	case "testHooksError":
		if res != nil && res.StatusCode != 400 {
			return nil, errors.New("expected status code 400")
		}
		return nil, errors.New("special test error case")
	case "statusGetDefaultError":
		if res != nil && res.StatusCode == 418 {
			return &http.Response{
				StatusCode: 200,
				Status:     "200 OK",
				Body:       http.NoBody,
				Header:     make(http.Header),
			}, nil
		}
	}

	return res, err
}

type TestClient struct {
	underlyingClient HTTPClient
}

var _ HTTPClient = (*TestClient)(nil)

func (c *TestClient) Do(req *http.Request) (*http.Response, error) {
	req.Header.Add("Client-Level-Header", "added by client")

	return c.underlyingClient.Do(req)
}
